File: 0658-find-k-closest-elements/0658-find-k-closest-elements.py
class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        start, end = 0, k - 1

        for i in range(k, len(arr)):
            if abs(arr[i] - x) < abs(arr[start] - x) or arr[i] == arr[start]:
                start += 1
                end += 1
            else:
                break
        
        return arr[start:end + 1]

File: 0658-find-k-closest-elements/0658-find-k-closest-elements.cpp
class Solution {
public:
    vector<int> findClosestElements(vector<int>& arr, int k, int x) {
        int lo = 0;
        int hi = arr.size() - k;
        int ans = 0;

        function<bool(int)> ok = [&] (int mid) {
            if (mid + k >= arr.size()) return false;
            
            if (x >= arr[mid + k]) return true;

            if (x <= arr[mid]) return false;

            return arr[mid + k] - x  < x - arr[mid];
        };

        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (ok(mid)) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
                ans = mid;
            }
        }

        vector<int> ansVec;
        for (int i = 0; i < k; i++) {
            ansVec.push_back(arr[ans + i]);
        }

        return ansVec;
    }
};